crypto/merkletrie/node.go

```diff
@@ -0,0 +1,383 @@
+
+package merkletrie
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/algorand/go-algorand/crypto"
+)
+
+type node struct {
+	hash         []byte
+	children     []storedNodeIdentifier
+	childrenNext []byte
+	firstChild   byte
+	leaf         bool
+}
+
+func (n *node) stats(cache *merkleTrieCache, stats *Stats, depth int) (err error) {
+	stats.nodesCount++
+	if n.leaf {
+		stats.leafCount++
+		if depth > stats.depth {
+			stats.depth = depth
+		}
+		stats.size += 4 + len(n.hash) + 1
+		return nil
+	}
+	i := n.firstChild
+	stats.size += 4 + len(n.hash) + cap(n.children)*8 + cap(n.childrenNext) + 1
+	for {
+		childNode, err := cache.getNode(n.children[i])
+		if err != nil {
+			return err
+		}
+		err = childNode.stats(cache, stats, depth+1)
+		if err != nil {
+			return err
+		}
+		if i == n.childrenNext[i] {
+			break
+		}
+		i = n.childrenNext[i]
+	}
+	return nil
+}
+
+// find searches the trie for the element, recursively.
+func (n *node) find(cache *merkleTrieCache, d []byte) (bool, error) {
+	if n.leaf {
+		return 0 == bytes.Compare(d, n.hash), nil
+	}
+	childNodeID := n.children[d[0]]
+	if childNodeID == storedNodeIdentifierNull {
+		return false, nil
+	}
+	childNode, err := cache.getNode(childNodeID)
+	if err != nil {
+		return false, err
+	}
+	return childNode.find(cache, d[1:])
+}
+
+// add adds an element to the sub-trie
+// assumption : we know that the key is absent from the tree
+func (n *node) add(cache *merkleTrieCache, d []byte, path []byte) (nodeID storedNodeIdentifier, err error) {
+	// allocate a new node to replace the current one.
+	var pnode *node
+	if n.leaf {
+		// find the diff index:
+		idiff := 0
+		for ; n.hash[idiff] == d[idiff]; idiff++ {
+		}
+
+		curChildNode, curChildNodeID := cache.allocateNewNode()
+		newChildNode, newChildNodeID := cache.allocateNewNode()
+
+		curChildNode.leaf = true
+		curChildNode.hash = n.hash[idiff+1:]
+		newChildNode.leaf = true
+		newChildNode.hash = d[idiff+1:]
+
+		pnode, nodeID = cache.allocateNewNode()
+		pnode.leaf = false
+		pnode.children = make([]storedNodeIdentifier, 256)
+		pnode.childrenNext = make([]byte, 256)
+
+		pnode.children[n.hash[idiff]] = curChildNodeID
+		pnode.children[d[idiff]] = newChildNodeID
+		if n.hash[idiff] < d[idiff] {
+			pnode.firstChild = n.hash[idiff]
+			pnode.childrenNext[pnode.firstChild] = d[idiff]
+			pnode.childrenNext[d[idiff]] = d[idiff]
+		} else {
+			pnode.firstChild = d[idiff]
+			pnode.childrenNext[pnode.firstChild] = n.hash[idiff]
+			pnode.childrenNext[n.hash[idiff]] = n.hash[idiff]
+		}
+		pnode.hash = append(path, d[:idiff]...)
+
+		for i := idiff - 1; i >= 0; i-- {
+			// create a parent node for pnode.
+			pnode2, nodeID2 := cache.allocateNewNode()
+			pnode2.leaf = false
+			pnode2.children = make([]storedNodeIdentifier, 256)
+			pnode2.childrenNext = make([]byte, 256)
+			pnode2.children[d[i]] = nodeID
+			pnode2.firstChild = d[i]
+			pnode2.childrenNext[d[i]] = d[i]
+			pnode2.hash = append(path, d[:i]...)
```

Comment:
If `path` has enough capacity, then all the `pnode2.hash` slices are going to have the same backing array. Is this intentional?

Reply:
(I don't see any immediate problems arising from this, but having multiple nodes whose `hash` fields all point to the same array seems likely to cause subtle problems if we ever write to that array or call `append(node.hash, x)` anywhere, so if it isn't intentional then we probably shouldn't do it)

Reply:
That's a good observation. I think that in the current usage it's advantageous.